inttest/nllb/nllb_test.go

```diff
@@ -0,0 +1,333 @@
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+	"sigs.k8s.io/yaml"
+)
+
+const kubeSystem = "kube-system"
+
+type suite struct {
+	common.FootlooseSuite
+}
+
+func (s *suite) TestNodeLocalLoadBalancing() {
+	const controllerArgs = "--kube-controller-manager-extra-args='--node-monitor-period=3s --node-monitor-grace-period=9s'"
+
+	ctx := s.Context()
+
+	{
+		config, err := yaml.Marshal(&v1beta1.ClusterConfig{
+			Spec: &v1beta1.ClusterSpec{
+				Network: func() *v1beta1.Network {
+					network := v1beta1.DefaultNetwork()
+					network.NodeLocalLoadBalancer.Enabled = true
+					return network
+				}(),
+
+				WorkerProfiles: v1beta1.WorkerProfiles{
+					v1beta1.WorkerProfile{
+						Name: "default",
+						Config: func() json.RawMessage {
+							kubeletConfig := kubeletv1beta1.KubeletConfiguration{
+								NodeStatusUpdateFrequency: metav1.Duration{Duration: 3 * time.Second},
+							}
+							bytes, err := json.Marshal(kubeletConfig)
+							s.Require().NoError(err)
+							return bytes
+						}(),
+					},
+				},
+			},
+		})
+		s.Require().NoError(err)
+
+		for i := 0; i < s.ControllerCount; i++ {
+			s.WriteFileContent(s.ControllerNode(i), "/tmp/k0s.yaml", config)
+		}
+	}
+
+	s.Require().True(s.Run("ControllerAndWorkersGetUp", func() {
+		s.Require().NoError(s.InitController(0, "--config=/tmp/k0s.yaml", controllerArgs))
+
+		s.T().Logf("Starting workers and waiting for cluster to become ready")
+
+		token, err := s.GetJoinToken("worker")
+		s.Require().NoError(err)
+		s.Require().NoError(s.RunWorkersWithToken(token))
+
+		clients, err := s.KubeClient(s.ControllerNode(0))
+		s.Require().NoError(err)
+
+		eg, _ := errgroup.WithContext(ctx)
+		for i := 0; i < s.WorkerCount; i++ {
+			nodeName := s.WorkerNode(i)
+			eg.Go(func() error {
+				if err := s.WaitForNodeReady(nodeName, clients); err != nil {
+					return fmt.Errorf("Node %s is not ready: %w", nodeName, err)
+				}
+				return nil
+			})
+		}
+		s.Require().NoError(eg.Wait())
+
+		s.Require().NoError(s.checkClusterReadiness(ctx, clients, 1))
+	}))
+
+	s.Require().True(s.Run("JoinNewControllers", func() {
+		token, err := s.GetJoinToken("controller")
+		s.Require().NoError(err)
+
+		eg, _ := errgroup.WithContext(ctx)
+		eg.Go(func() error { return s.InitController(1, "--config=/tmp/k0s.yaml", controllerArgs, token) })
+		eg.Go(func() error { return s.InitController(2, "--config=/tmp/k0s.yaml", controllerArgs, token) })
+
+		s.Require().NoError(eg.Wait())
+
+		clients, err := s.KubeClient(s.ControllerNode(1))
+		s.Require().NoError(err)
+
+		s.T().Logf("Checking if HA cluster is ready")
+		s.Require().NoError(s.checkClusterReadiness(ctx, clients, s.ControllerCount))
+	}))
+
+	workerNameToRestart := s.WorkerNode(0)
+	s.Require().True(s.Run("CycleControllers", func() {
+		for i := 0; i < s.ControllerCount; i++ {
+			controllerName := s.ControllerNode(i)
+			s.Require().True(s.Run(controllerName, func() {
+				s.Require().True(s.Run("StopWorker", func() {
```

Comment:
Why all the nesting of `s.Require.True(s.Run(...`? I do get the "flow" of the test case here but I do not get the "nesting" part.

Reply:
I have the question but not about nested .Run. 

What is the reason to have explicit checks for the result t.Run?

Reply:
> What is the reason to have explicit checks for the result t.Run?

This is to be fail-fast and stop the parent test when the first subtest fails. Otherwise, the suite would continue with the next subtest, which would be unlikely to succeed given that its predecessor(s) failed. I found this to be the least cumbersome solution in conjunction with subtests. An alternative would be to not use subtests at all and rely more on s.T().Log() to report progress. Personally, I find the subtest output a bit more appealing, and it also gives timings for each step out of the box.

Reply:
We use Require in those sub-tests.
According to the documentation Require should fail fast already. 

Doesn't it work as expected?

Reply:
The code currently uses `s.Run(string, func())` instead of `s.T().Run(string, func(*testing.T))`. The former adjusts the suite's test pointer to the subtest as long as the subtest function is being executed, hence eliding the additional `*testing.T` from that function's signature. So all assertions via `s` called from within the subtest function will operate on the subtest, not the parent test.

That behavior can be bypassed by using `s.T().Run(string, func(*testing.T))` instead and not using the subtest's `t`. This gives weirder errors and log statements, though. Because the subtest is never used, everything operates on the parent test, making the subtest rather obsolete. Especially calling `FailNow` (that's what testify's require does) on a parent test from within a subtest always gives messages like "testing.go:1343: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test".

To clarify the current behavior, it could be beneficial to introduce a helper function on the suite, like so:

```go
func (s *Suite) runOrFailNow(name string, f func()) {
	if !s.Run(name, f) {
		s.T().FailNow()
	}
}
```

Then the current

```go
s.Require().True(s.Run("subtest", func() {
	// ...
}))
```

would become

```go
s.runOrFailNow("subtest", func() {
	// ...
})
```

Removing the subtests alltogether remains another option: adding more `s.T().Logf(...)` calls or augmenting existing log statements with additional context.

Reply:
what if you use the parent t in the subtest instead of its own t? for example by naming the argument as _ to not overshadow the parent one?

Reply:
> what if you use the parent t in the subtest instead of its own t? for example by naming the argument as _ to not overshadow the parent one?

Using the parent test from within the subtest yields to the stuff I tried to describe here:

> This gives weirder errors and log statements, though. [...] Especially calling FailNow (that's what testify's require does) on a parent test from within a subtest always gives messages like "testing.go:1343: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test".

For example, testify errors will reference the parent test instead of the subtest as the failed test, and when using t.Log, this will always look like below, giving the impression that the subtest would do nothing, or run silently in parallel:

```text
=== RUN   parent/subtest
=== CONT  parent
    ...
```

Moreover, having references to both the parent test and the subtest around will always lead to the question of which one to use. Logging could be done via subtests, but assertions via the parent? This is going to get complicated pretty quickly.

A final option besides the helper method or the omission of subtests would be to just not be fail-fast and let subsequent subtests run and have them fail as well. This will be noisier when test failures actually occur, as each subsequent subtest will report its own failures, but should be more or less fine otherwise. (There might be some problems during test timeouts, though.)

Reply:
>A final option besides the helper method or the omission of subtests would be to just not be fail-fast and let subsequent subtests run and have them fail as well

I vote for this. 

Another option to consider is instead of t.Run use s.Run which seems to be merged with that PR https://github.com/stretchr/testify/pull/655/files

>Adds a .Run(..) method onto test suites, mimicking the semantics of [testing.Run](https://godoc.org/testing#T.Run). This will make writing table-driven tests in test suites cleaner and simpler, by handling the management of *testing.T instances transparently so that sub-tests can use the same suite.Assert() and suite.Require() methods, and have all output and test logic properly fall under the correct subtest.

@jnummelin @makhov what's your opinion here, folks?

Reply:
Oh we are using s.Run already... I wonder why it doesn't fall as expected... Well, then I vote for not having fail-fast approach now.

Reply:
I don't understand, what is the problem here. `require` fails the current test and I'd say it's exactly the expected behavior. I do want to know which tests are failing and which are passing.

> This is to be fail-fast and stop the parent test when the first subtest fails

I don't see why we need to be fail-fast in unit tests. But if we need to, it may indicate the misuse of subtests

Reply:
> I wonder why it doesn't fall as expected

I think it behaves as intended. It returns the outcome of the subtest as a bool, just the same way as `t.Run` does it. So we need to look at that bool ourselves.

>  vote for not having fail-fast approach now.

Will do.

Reply:
Done. Flattened the test structure and removed the "fail-fastness".